File: java/src/acme/infrastructure/inmemory/domain/InMemoryUsers.java
package acme.infrastructure.inmemory.domain;

import acme.domain.Users;
import acme.domain.model.User;

import java.util.HashMap;
import java.util.Map;

public class InMemoryUsers implements Users {
    private Map<String, User> users;

    public InMemoryUsers() {
        users = new HashMap<>();
    }

    @Override
    public void add(User user) {
        users.put(user.email(), user);
    }

    @Override
    public boolean exists(String email) {
        return users.containsKey(email);
    }

    @Override
    public User getByEmail(String email) {
        return users.get(email);
    }

    @Override
    public void remove(String email) {
        users.remove(email);
    }
}

File: kotlin/src/acme/infrastructure/inmemory/domain/InMemoryUsers.kt
package acme.infrastructure.inmemory.domain

import acme.domain.Users
import acme.domain.model.User

import java.util.HashMap

class InMemoryUsers : Users {
    private val users: MutableMap<String, User>

    init {
        users = HashMap()
    }

    override fun add(user: User) {
        users[user.email()] = user
    }

    override fun exists(email: String): Boolean {
        return users.containsKey(email)
    }

    override fun getByEmail(email: String): User? {
        return users[email]
    }

    override fun remove(email: String) {
        users.remove(email)
    }
}
